# Others

## Function One：TE_classification_and_counting

Categorize and count the results of TE.

```
# default
python te_classification_and_counting.py GWHBCHF00000000.genome.fasta.out

# "Optional parameter -o output_path"
python te_classification_and_counting.py GWHBCHF00000000.genome.fasta.out -o ./
```

## Function Two：interproscan_classification_and_counting

Categorize and count the results of interproscan.

```
python interproscan_classification_and_counting.py interproscan.result.tsv interproscan.result.count.txt
```

## Function Three：WGDI file to JCVI file

Generate input files for JCVI using the input or output results of WGDI

Each function and usage can view help information using the `-h` parameter.

```
usage: wgdi2jcvi.py [-h] {blast2anchor,gff2bed,block2simple} ...

Process collinearity file, convert GFF to BED, or generate anchors from BLAST results.

positional arguments:
  {blast2anchor,gff2bed,block2simple}
    blast2anchor        Generate anchor file based on collinearity and BLAST results.
    gff2bed             Convert GFF file to BED format.
    block2simple        Process collinearity file and output a simple summary.
```

### Usage 1 :wgdi gff to jcvi bed
```                           
usage: wgdi2jcvi.py gff2bed [-h] -gff input_gff_file -bed output_bed_file

options:
  -h, --help            show this help message and exit
  -gff input_gff_file   Path to the input GFF file.
  -bed output_bed_file  Path to the output BED file.
```
```
# example
python wgdi2jcvi.py gff2bed -gff coca_gff -bed coca.bed
```

### Usage 2 :blast to jcvi anchor
```
usage: wgdi2jcvi.py blast2anchor [-h] -c collinearity_file -b blast_file -o output_anchor_file

options:
  -h, --help            show this help message and exit
  -c collinearity_file  Path to the input collinearity file generated by WGDI.
  -b blast_file         Path to the BLAST result file.
  -o output_anchor_file
                        Path to the output anchor file.
```
```
# example
python wgdi2jcvi.py blast2anchor -c coca_schi.collinearity.txt -b coca_schi.1e-5.blast -o coca_schi.anchor   
```
### Usage 3 :blast to jcvi anchor
```
python wgdi2jcvi.py block2simple -h
usage: wgdi2jcvi.py block2simple [-h] -c input_collinearity_file -o output_simple_file [--filter filter_blockid_file]

options:
  -h, --help            show this help message and exit
  -c input_collinearity_file
                        Path to the input collinearity file generated by WGDI.
  -o output_simple_file
                        Path to the output simple file.
  --filter filter_blockid_file
                        [option] Path to the filter block ID file.
```
```
# example
python wgdi2jcvi.py block2simple -c coca_schi.collinearity.txt -o coca_schi.simple --filter coca_schi.filter_blockid.txt

```


